File: Python3/0393_utf_8_validation.py
# https://leetcode.cn/problems/utf-8-validation/

class Solution:
    def validUtf8(self, data):
        """
        Inputs:
            data: list[int]
        Outputs:
            res: bool
        """

        i = 0
        n = len(data)
        while i < n:
            num = data[i]
            if (num & 128) == 0:
                i = i + 1
                continue

            j = 0
            while j < 4:
                if (num & 128) != 0:
                    j = j + 1
                    num = num << 1
                else:
                    break
            if (j == 1) or ((num & 128) != 0):
                return False

            for k in range(1, j):
                if i + k >= n:
                    return False
                num = data[i+k]
                if (num & 128) == 0:
                    return False
                num = num << 1
                if (num & 128) != 0:
                    return False

            i = i + j
        return True

File: Cpp/0393_utf_8_validation.cpp
// https://leetcode.cn/problems/utf-8-validation/

class Solution {
public:
    bool validUtf8(vector<int>& data) {
        int i = 0, n = data.size();
        while (i < n) {
            int num = data[i];
            if ((num & 128) == 0) {
                i = i + 1;
                continue;
            }

            int j = 0;
            while (j < 4) {
                if ((num & 128) != 0) {
                    j = j + 1;
                    num = num << 1;
                }
                else {
                    break;
                }
            }
            if ((j == 1) || ((num & 128) != 0)) {
                return false;
            }

            for (int k = 1; k < j; ++k) {
                if (i + k >= n) {
                    return false;
                }
                num = data[i + k];
                if ((num & 128) == 0) {
                    return false;
                }
                num = num << 1;
                if ((num & 128) != 0) {
                    return false;
                }
            }
            i = i + j;
        }
        return true;
    }
};
